### PR TITLE
821 series diff timestamp2

### LINKED ISSUE #821: `Series[Timestamp].diff()` should return `Series[Timedelta]`
**Describe the bug**
`Series[Timestamp].diff()` returns `Series[Timedelta]` but stubs indicate it's `Series[Timestamp]` instead.

**To Reproduce**
```python
# python bug.py & mypy bug.py & pyright bug.py
from __future__ import annotations

from typing import TYPE_CHECKING, reveal_type

import pandas as pd

if TYPE_CHECKING:
    from pandas.core.series import TimedeltaSeries, TimestampSeries

def fun1(times: pd.Series[pd.Timestamp]) -> pd.Series[pd.Timedelta]:
    reveal_type(times)
    reveal_type(times.diff())
    print(type(times[1]))
    print(type(times.diff()[1]))
    return times.diff()

def fun2(times: TimestampSeries) -> TimedeltaSeries:
    reveal_type(times)
    reveal_type(times.diff())
    print(type(times[1]))
    print(type(times.diff()[1]))
    return times.diff()

times = pd.Series([pd.Timestamp(0), pd.Timestamp(0)])
fun1(times)
fun2(times)
```

Output:
```
Runtime type is 'Series'
Runtime type is 'Series'
<class 'pandas._libs.tslibs.timestamps.Timestamp'>
<class 'pandas._libs.tslibs.timedeltas.Timedelta'>
Runtime type is 'Series'
Runtime type is 'Series'
<class 'pandas._libs.tslibs.timestamps.Timestamp'>
<class 'pandas._libs.tslibs.timedeltas.Timedelta'>
bug.py:13: note: Revealed type is "pandas.core.series.Series[pandas._libs.tslibs.timestamps.Timestamp]"
bug.py:14: note: Revealed type is "pandas.core.series.Series[pandas._libs.tslibs.timestamps.Timestamp]"
bug.py:17: error: Incompatible return value type (got "Series[Timestamp]", expected "Series[Timedelta]")  [return-value]
bug.py:21: note: Revealed type is "pandas.core.series.TimestampSeries"
bug.py:22: note: Revealed type is "pandas.core.series.Series[pandas._libs.tslibs.timestamps.Timestamp]"
bug.py:25: error: Incompatible return value type (got "Series[Timestamp]", expected "TimedeltaSeries")  [return-value]
Found 2 errors in 1 file (checked 1 source file)
C:\Build\project\bug.py
  C:\Build\project\bug.py:13:17 - information: Type of "times" is "Series[Timestamp]"
  C:\Build\project\bug.py:14:17 - information: Type of "times.diff()" is "Series[Timestamp]"
  C:\Build\project\bug.py:17:12 - error: Expression of type "Series[Timestamp]" cannot be assigned to return type "Series[Timedelta]"
    "Series[Timestamp]" is incompatible with "Series[Timedelta]"
      Type parameter "S1@Series" is invariant, but "Timestamp" is not the same as "Timedelta" (reportGeneralTypeIssues)
  C:\Build\project\bug.py:21:17 - information: Type of "times" is "TimestampSeries"
  C:\Build\project\bug.py:22:17 - information: Type of "times.diff()" is "Series[Timestamp]"
  C:\Build\project\bug.py:25:12 - error: Expression of type "Series[Timestamp]" cannot be assigned to return type "TimedeltaSeries"
    "Series[Timestamp]" is incompatible with "TimedeltaSeries" (reportGeneralTypeIssues)
2 errors, 0 warnings, 4 informations
```

**Please complete the following information:**
- Windows 10 22H2
 - Python 3.12.0
 - `mypy==1.7.1`, `pyright==1.1.337`, `pandas-stubs==2.1.1.230928`

**Additional context**
I tried both `Series[Timestamp]` and `TimestampSeries` due to https://github.com/pandas-dev/pandas-stubs/issues/673#issuecomment-1739434074
Possibly related: https://github.com/pandas-dev/pandas-stubs/issues/718

### PR DESCRIPTION
**Pull request by mutricyl**
- [x] Closes #821 
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

- check function updated for series.diff particular usage
- it looks like only first `assert_never` is actually checked so I created several test functions

### CODE REVIEW

File: pandas-stubs/core/series.pyi [hunk 1]
```diff
@@ -2168,11 +2181,13 @@ class TimedeltaSeries(Series[Timedelta]):
         numeric_only: _bool = ...,
         **kwargs,
     ) -> Timedelta: ...
+    def diff(self, periods: int = ...) -> TimedeltaSeries: ...  # type: ignore[override] # pyright: ignore[reportIncompatibleMethodOverride]
 
 class PeriodSeries(Series[Period]):
     @property
     def dt(self) -> PeriodProperties: ...  # type: ignore[override] # pyright: ignore[reportIncompatibleMethodOverride]
     def __sub__(self, other: PeriodSeries) -> OffsetSeries: ...  # type: ignore[override] # pyright: ignore[reportIncompatibleMethodOverride]
+    def diff(self, periods: int = ...) -> Series[type[object]]: ...  # type: ignore[override] # pyright: ignore[reportIncompatibleMethodOverride]
```
Comment (Dr-Irv) [inline pandas-stubs/core/series.pyi:2190]: I think this should be `OffsetSeries` as the return type.

File: tests/test_series.py [hunk 2]
```diff
@@ -3069,3 +3070,103 @@ def first_arg_not_series(argument_1: int, ser: pd.Series) -> pd.Series:
...
+            pd.Series([pd.Timedelta(0), pd.Timedelta(1)]).diff(), "TimedeltaSeries"
+        ),
+        pd.Series,
+        pd.Timedelta,
+        index_to_check_for_type=-1,
+    )
+    # period -> object
+    with pytest_warns_bounded(
+        RuntimeWarning, "overflow encountered in scalar multiply"
+    ):
```
Comment (Dr-Irv) [inline tests/test_series.py:3115]: this shouldn't be necessary

File: tests/test_series.py [hunk 3]
```diff
@@ -3069,3 +3070,103 @@ def first_arg_not_series(argument_1: int, ser: pd.Series) -> pd.Series:
...
+    # object -> object
+    check(
+        assert_type(s.astype(object).diff(), "pd.Series[type[object]]"),
+        pd.Series,
+        object,
+    )
+    # Baseoffset (object) -> object
+    # from pandas.tseries.frequencies import to_offset
+    # pd.Series([to_offset("5min"), to_offset("1D1h")]).diff()
+    # pd.Series([pd.DateOffset(days=1), pd.DateOffset(days=2)]).diff()
```
Comment (Dr-Irv) [inline tests/test_series.py:3144]: Not sure why this is there?

File: tests/test_series.py [hunk 4]
```diff
@@ -3069,3 +3070,103 @@ def first_arg_not_series(argument_1: int, ser: pd.Series) -> pd.Series:
...
+    # complex -> complex
+    check(
+        assert_type(s.astype(complex).diff(), "pd.Series[complex]"), pd.Series, complex
+    )
+    if TYPE_CHECKING_INVALID_USAGE:
+        # interval -> TypeError: IntervalArray has no 'diff' method. Convert to a suitable dtype prior to calling 'diff'.
+        assert_never(pd.Series([pd.Interval(0, 2), pd.Interval(1, 4)]).diff())
+        # datetime.time -> TypeError: unsupported operand type(s) for -: 'datetime.time' and 'datetime.time'
+        # pd.Series([datetime.datetime.now().time(), datetime.datetime.now().time()]).diff()
```
Comment (Dr-Irv) [inline tests/test_series.py:3153]: If you remove the `.time()` calls, then this is valid code.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
You'll need to merge with `main` so that the tests will run.  Some small things I picked up.

File: tests/test_series.py (hunk 2 above)
Comment (mutricyl) [reply tests/test_series.py:3115]: it is for pytest to run successfully : when running diff of Series[Period] a RuntimeWarning is raised. I guess it is close to be a pandas core bug (or maybe numpy)

Comment (Dr-Irv) [reply tests/test_series.py:3115]: I think that's a pandas bug.  So can you change your test to:
```python
pd.Series(pd.period_range(start='2017-01-01', end='2017-02-01', freq='D')).d
iff()
```
That won't show the bug

File: tests/test_series.py (hunk 3 above)
Comment (mutricyl) [reply tests/test_series.py:3144]: sorry I left some mess around. I'll clean it

File: tests/test_series.py (hunk 2 above)
Comment (mutricyl) [reply tests/test_series.py:3115]: still have the same issue with `pd.period_range`

``` python
___________________________________________________________________________________________________ test_diff ____________________________________________________________________________________________________ 

    def test_diff() -> None:
        s = pd.Series([1, 1, 2, 3, 5, 8])
        # int -> float
        check(assert_type(s.diff(), "pd.Series[float]"), pd.Series, float)
        # unint -> float
        check(assert_type(s.astype(np.uint32).diff(), "pd.Series[float]"), pd.Series, float)
        # float -> float
        check(assert_type(s.astype(float).diff(), "pd.Series[float]"), pd.Series, float)
        # datetime.date -> timeDelta
        check(
            assert_type(
                pd.Series(
                    [datetime.datetime.now().date(), datetime.datetime.now().date()]
                ).diff(),
                "TimedeltaSeries",
            ),
            pd.Series,
            pd.Timedelta,
            index_to_check_for_type=-1,
        )
        # timestamp -> timedelta
        times = pd.Series([pd.Timestamp(0), pd.Timestamp(1)])
        check(
            assert_type(times.diff(), "TimedeltaSeries"),
            pd.Series,
            pd.Timedelta,
            index_to_check_for_type=-1,
        )
        # timedelta -> timedelta64
        check(
            assert_type(
                pd.Series([pd.Timedelta(0), pd.Timedelta(1)]).diff(), "TimedeltaSeries"
            ),
            pd.Series,
            pd.Timedelta,
            index_to_check_for_type=-1,
        )
        # period -> object
        check(
            assert_type(
                pd.Series(
                    pd.period_range(start="2017-01-01", end="2017-02-01", freq="D")
>               ).diff(),
                "pd.Series[type[object]]",
            ),
            pd.Series,
            object,
        )

tests\test_series.py:3117:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
C:\Users\laurent.mutricy\AppData\Local\miniconda3\envs\pandas-stubs\Lib\site-packages\pandas\core\series.py:3111: in diff
    result = algorithms.diff(self._values, periods)
C:\Users\laurent.mutricy\AppData\Local\miniconda3\envs\pandas-stubs\Lib\site-packages\pandas\core\algorithms.py:1379: in diff
    return op(arr, arr.shift(n))
C:\Users\laurent.mutricy\AppData\Local\miniconda3\envs\pandas-stubs\Lib\site-packages\pandas\core\ops\common.py:76: in new_method
    return method(self, other)
C:\Users\laurent.mutricy\AppData\Local\miniconda3\envs\pandas-stubs\Lib\site-packages\pandas\core\arrays\datetimelike.py:1460: in __sub__
    result = self._sub_periodlike(other)
C:\Users\laurent.mutricy\AppData\Local\miniconda3\envs\pandas-stubs\Lib\site-packages\pandas\core\arrays\datetimelike.py:1304: in _sub_periodlike
    new_data = np.array([self.freq.base * x for x in new_i8_data])
C:\Users\laurent.mutricy\AppData\Local\miniconda3\envs\pandas-stubs\Lib\site-packages\pandas\core\arrays\datetimelike.py:1304: in <listcomp>
    new_data = np.array([self.freq.base * x for x in new_i8_data])
offsets.pyx:1033: in pandas._libs.tslibs.offsets.Tick.__mul__
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

>   ???
E   RuntimeWarning: overflow encountered in scalar multiply

offsets.pyx:487: RuntimeWarning
```

File: tests/test_series.py [hunk 5]
```diff
@@ -3124,13 +3133,25 @@ def test_diff() -> None:
...
-    ):
+    if WINDOWS:
+        with pytest_warns_bounded(
+            RuntimeWarning, "overflow encountered in scalar multiply"
+        ):
+            check(
+                assert_type(
+                    pd.Series(
+                        pd.period_range(start="2017-01-01", end="2017-02-01", freq="D")
+                    ).diff(),
```
Comment (Dr-Irv) [inline tests/test_series.py:3144]: If you use `pd.Series(pd.period_range(start="10:00", end="18:00", freq="h")).diff()`, then we can avoid the Windows issue.  Can you use that test instead?

File: tests/test_series.py [hunk 6]
```diff
@@ -3124,13 +3133,25 @@ def test_diff() -> None:
...
+                    "pd.Series[type[object]]",
```
Comment (Dr-Irv) [inline tests/test_series.py:3145]:
```suggestion
                    "OffsetSeries"
```

File: tests/test_series.py [hunk 7]
```diff
@@ -3152,19 +3173,17 @@ def test_diff() -> None:
...
     )
     if TYPE_CHECKING_INVALID_USAGE:
         # interval -> TypeError: IntervalArray has no 'diff' method. Convert to a suitable dtype prior to calling 'diff'.
         assert_never(pd.Series([pd.Interval(0, 2), pd.Interval(1, 4)]).diff())
-        # datetime.time -> TypeError: unsupported operand type(s) for -: 'datetime.time' and 'datetime.time'
-        # pd.Series([datetime.datetime.now().time(), datetime.datetime.now().time()]).diff()
+
+    # note managed by stubs:
+    # datetime.time -> Never (TypeError: unsupported operand type(s) for -: 'datetime.time' and 'datetime.time')
+    # Baseoffset (object) -> object
```
Comment (Dr-Irv) [inline tests/test_series.py:3186]: remove this comment

File: pandas-stubs/core/series.pyi (hunk 1 above)
Comment (Dr-Irv) [reply pandas-stubs/core/series.pyi:2193]: `OffsetSeries` is only in the type stubs.  Please use that.

File: tests/test_series.py [hunk 8]
```diff
@@ -3091,3 +3093,113 @@ def test_series_apply() -> None:
...
+        ):
+            check(
+                assert_type(
+                    pd.Series(
+                        pd.period_range(start="2017-01-01", end="2017-02-01", freq="D")
+                    ).diff(),
+                    "pd.Series[type[object]]",
+                ),
+                pd.Series,
+                object,
```
Comment (Dr-Irv) [inline tests/test_series.py:3148]:
```suggestion
                BaseOffset,
                index_to_check_for_type=-1
```

File: tests/test_series.py (hunk 2 above)
Comment (Dr-Irv) [reply tests/test_series.py]: It's Windows only when the frequency is "D".  So if we use "h", we don't get the warning.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

File: tests/test_series.py (hunk 5 above)
Comment (Dr-Irv) [reply tests/test_series.py:3144]: Never mind.  Even with `freq="h"`, the same issue occurs.

Comment (mutricyl) [reply tests/test_series.py:3144]: do you prefer to keep this test as is (with the pytest_warns_bounded) or shall I just remove it ?

Comment (Dr-Irv) [reply tests/test_series.py:3147]: keep as is

**Comment by Dr-Irv** [comment]
@mutricyl I'm fine with all of your changes, but the CI is failing on MacOS.  It has something to do with tables.

Can you try updating `pyproject.toml` to refer to `tables` version 3.9.2 to see if that will fix the problem?

**Comment by mutricyl** [comment]
tables 3.9.2 does not solve the issue apparently

**Comment by Dr-Irv** [comment]
> tables 3.9.2 does not solve the issue apparently

Ugh.  We'll have to triage that against previous PR's that were working fine.  

@twoertwein any ideas here?  MacOS CI is failing because it can't build `tables`.  I'm guessing this might be an architecture problem.  Maybe we should lock MacOS to x86?

**Comment by Dr-Irv** [comment]
> > tables 3.9.2 does not solve the issue apparently
> 
> Ugh. We'll have to triage that against previous PR's that were working fine.
> 
> @twoertwein any ideas here? MacOS CI is failing because it can't build `tables`. I'm guessing this might be an architecture problem. Maybe we should lock MacOS to x86?

I fixed the issue - changed `macos-latest` to `macos-13` and it worked.

@mutricyl can you merge with the latest `main` and push ?

**Comment by mutricyl** [comment]
@Dr-Irv , I'am done

**Review by Dr-Irv — APPROVED** [review]
thanks @mutricyl

**Pull request merged**